biggus/__init__.py

```diff
@@ -839,10 +891,114 @@ def _getitem_full_keys(self, keys):
             sub_array = self.array
         else:
             sub_array = self.array[tuple(array_keys)]
-        return BroadcastArray(sub_array, new_broadcast_dict)
+        return BroadcastArray(sub_array, new_broadcast_dict, leading_shape)
+
+    @classmethod
+    def broadcast_arrays(cls, array1, array2):
+        """
+        Given two arrays, if broadcasting needs to take place on either arrays
+        to make their shapes comparable, create BroadcastArrays for each.
+
+        """
+        shape, bcast_kwargs1, bcast_kwargs2 = (
+            BroadcastArray.compute_broadcast_kwargs(array1.shape,
+                                                    array2.shape))
+        if bcast_kwargs1['broadcast'] or bcast_kwargs1['leading_shape']:
+            array1 = BroadcastArray(array1, **bcast_kwargs1)
+
+        if bcast_kwargs2['broadcast'] or bcast_kwargs2['leading_shape']:
+            array2 = BroadcastArray(array2, **bcast_kwargs2)
+
+        return array1, array2
+
+    @staticmethod
+    def compute_broadcast_shape(shape1, shape2):
+        """
+        Given two shapes, use numpy's broadcasting rules to compute the
+        broadcasted shape.
+
+        """
+        # Rule 1: If the two arrays differ in their number of dimensions, the
+        # shape of the array with fewer dimensions is padded with ones on its
+        # leading (left) side.
+        s1, s2 = list(shape1), list(shape2)
+        len_diff = len(s1) - len(s2)
+        if len_diff > 0:
+            s2[0:0] = [1] * len_diff
+        else:
+            s1[0:0] = [1] * -len_diff
+
+        # Rule 2: If the shape of the two arrays does not match in any
+        # dimension, the array with shape equal to 1 in that dimension is
+        # stretched to match the other shape.
+        shape = []
+        for size1, size2 in zip(s1, s2):
+            if size1 == size2:
+                shape.append(size1)
+            elif size1 == 1:
+                shape.append(size2)
+            elif size2 == 1:
+                shape.append(size1)
+            else:
+                # Rule 3: If in any dimension the sizes disagree and neither is
+                # equal to 1, an error is raised.
+                raise ValueError('operands could not be broadcast together '
+                                 'with shapes ({}) ({})'
+                                 ''.format(','.join(map(str, shape1)),
+                                           ','.join(map(str, shape2))))
+        return tuple(shape)
+
+    @classmethod
+    def compute_broadcast_kwargs(cls, shape1, shape2):
```

Comment:
Make this (and all the other class/static methods except `broadcast_arrays()`) private?